ENIGMAsystem/SHELL/Graphics_Systems/General/GSvertex.cpp

```diff
@@ -118,25 +104,77 @@ unsigned vertex_get_number(int buffer) {
-  if (enigma::vertexBuffers[buffer]->frozen) return;
-  enigma::vertexBuffers[buffer]->frozen = true;
-  enigma::vertexBuffers[buffer]->dirty = true;
+  if (vertexBuffer->frozen) return;
+  vertexBuffer->frozen = true;
+  vertexBuffer->dirty = true;
 }
 
-void vertex_submit(int buffer, int primitive) {
-  vertex_submit(buffer, primitive, 0, vertex_get_number(buffer));
+void vertex_clear(int buffer) {
+  enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+
+  // clear it just for good measure, even though it's probably already empty
+  // since we do that just after uploading it to the GPU "peer"
+  vertexBuffer->vertices.clear();
+  vertexBuffer->number = 0;
+
+  // we can clear a.k.a. "unfreeze" the vertex buffer only if it is actually frozen
+  if (!vertexBuffer->frozen) return;
+  vertexBuffer->frozen = false;
+  vertexBuffer->dirty = true;
 }
 
-void vertex_submit(int buffer, int primitive, int texture) {
-  texture_set(texture);
-  vertex_submit(buffer, primitive);
+void vertex_begin(int buffer, int format) {
+  enigma::vertexBuffers[buffer]->vertices.clear();
+  enigma::vertexBuffers[buffer]->format = format;
 }
 
-void vertex_submit(int buffer, int primitive, int texture, unsigned vertex_start, unsigned vertex_count) {
-  texture_set(texture);
-  vertex_submit(buffer, primitive, vertex_start, vertex_count);
+void vertex_end(int buffer) {
+  enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+
+  // we can only flag the vertex buffer contents as dirty and needing an update
+  // if the vertex buffer hasn't been frozen, otherwise we just ignore it
+  if (vertexBuffer->frozen) return;
+  vertexBuffer->dirty = true;
+  vertexBuffer->number = vertexBuffer->vertices.size();
+}
+
+void vertex_data(int buffer, const enigma::varargs& data) {
+  enigma::VertexBuffer* vertexBuffer = enigma::vertexBuffers[buffer];
+  const enigma::VertexFormat* vertexFormat = enigma::vertexFormats[vertexBuffer->format];
+  int attrIndex = 0;
+  for (int i = 0; i < data.argc;) {
+    attrIndex = attrIndex % vertexFormat->flags.size();
+    std::pair<int,int> attr = vertexFormat->flags[attrIndex++];
+    switch (attr.first) {
+      case vertex_type_float1:
+        vertex_float1(buffer, data.get(i++));
+        continue;
+      case vertex_type_float2:
+        vertex_float2(buffer, data.get(i), data.get(i + 1));
+        i += 2;
+        continue;
+      case vertex_type_float3:
+        vertex_float3(buffer, data.get(i), data.get(i + 1), data.get(i + 2));
+        i += 3;
+        continue;
+      case vertex_type_float4:
+        vertex_float4(buffer, data.get(i), data.get(i + 1), data.get(i + 2), data.get(i + 3));
+        i += 4;
+        continue;
+      case vertex_type_color:
+        vertex_argb(buffer, data.get(i++));
+        continue;
+      case vertex_type_ubyte4:
+        vertex_ubyte4(buffer, data.get(i), data.get(i + 1), data.get(i + 2), data.get(i + 3));
+        i += 4;
+        continue;
+    }
+    break;
```

Comment:
This is an error case, right? Might be useful to report it.

Reply:
Yeah, that's definitely an error case, should I only report it in debug mode or? Not sure what to do, I can just always print, but I don't know if I should print only in debug mode.

Reply:
No preference. It doesn't cost anything to report it in normal mode, since this is already a switch statement.

Reply:
Done: 41acab9cb1ccd0f95df123881eeadfa4529312dd